Tests/MongoSwiftSyncTests/SpecTestRunner/SpecTest.swift

```diff
@@ -257,37 +328,70 @@ extension SpecTest {
 
         print("Executing test: \(self.description)")
 
-        let clientOptions = self.clientOptions ?? ClientOptions(retryReads: true)
+        let clientOptions = self.clientOptions?.toClientOptions()
+
+        let client = try MongoClient.makeTestClient(options: clientOptions)
+        let monitor = client.addCommandMonitor()
+
+        if let collName = collName {
+            _ = try? client.db(dbName).createCollection(collName)
+            // Run the distinct command before every test to prevent `StableDbVersion` error in sharded cluster
+            // transactions. This workaround can be removed once SERVER-39704 is resolved.
+            _ = try? client.db(dbName).collection(collName).distinct(fieldName: "_id")
+        }
 
         if let failPoint = self.failPoint {
             try parent.activateFailPoint(failPoint)
         }
         defer { parent.disableActiveFailPoint() }
 
-        let client = try MongoClient.makeTestClient(options: clientOptions)
-        let monitor = client.addCommandMonitor()
-
-        let db = client.db(dbName)
-        var collection: MongoCollection<Document>?
-
-        if let collName = collName {
-            collection = db.collection(collName)
+        var sessions = [String: ClientSession]()
+        for session in Self.sessionNames {
+            sessions[session] = client.startSession(options: self.sessionOptions?[session])
         }
 
+        var sessionIds = [String: Document]()
+
         try monitor.captureEvents {
             for operation in self.operations {
                 try operation.validateExecution(
                     client: client,
-                    database: db,
-                    collection: collection,
-                    session: nil
+                    dbName: dbName,
+                    collName: collName,
+                    sessions: sessions
                 )
             }
+            // Keep track of the session IDs assigned to each session.
+            // Deinitialize each session thereby implicitly ending them.
+            for session in sessions.keys {
+                sessionIds[session] = sessions[session]?.id
+                sessions[session] = nil
+            }
+        }
+
+        let events = monitor.commandStartedEvents().map { commandStartedEvent in
+            TestCommandStartedEvent(from: commandStartedEvent, sessionIds: sessionIds)
         }
-        let events = monitor.commandStartedEvents().map { TestCommandStartedEvent(from: $0) }
 
         if let expectations = self.expectations {
             expect(events).to(match(expectations), description: self.description)
         }
+
+        try self.checkOutcome(dbName: dbName, collName: collName)
+    }
+
+    internal func checkOutcome(dbName: String, collName: String?) throws {
+        if let outcome = self.outcome {
```

Comment:
because this methods entire body is contained by this `if let`, we can just make it a `guard let` that returns on failure instead.

Reply:
Done.